File: StroidCfg/src/StroidCfg.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <unistd.h>
#include <sys/types.h>
#include <pwd.h>
#include <cassert>
#include "json.hh"
using namespace std;
using namespace JSON;

JSON::Value configInit()
{
    //get home directory
    int myuid;
    passwd *homedir;
    myuid = getuid();
    homedir = getpwuid(myuid);
    string path = homedir->pw_dir;
    path += "/StroidConfig.json";

    ofstream cfgFile;

    //create file if it does not exist
    if (!ifstream(path.c_str()))
    {
        cfgFile.open(path.c_str());
        cerr << "StroidConfig.json file does not exist... creating." << endl;
        cfgFile.close();
    }

    //if the file does exist, check its values
    Value in = parse_file(path.c_str());
    Object out = in;

    //all if statements make sure that the property exists
    //and that the type is correct
    if (!in["REF_HZ"] || in["REF_HZ"].type() == NIL
        || in["REF_HZ"].type() == BOOL)
        out["REF_HZ"] = 120;

    if (in["SHIP"].type() == OBJECT)
    {
        if (!in["SHIP"]["SPEED"] || in["SHIP"]["SPEED"].type() != INT
            || in["SHIP"]["SPEED"].type() != FLOAT)
            out["SHIP"]["SPEED"] = 5.01;

        if (!in["SHIP"]["FWD_THRUST"] || in["SHIP"]["FWD_THRUST"].type() != INT
            || in["SHIP"]["FWD_THRUST"].type() != FLOAT)
            out["SHIP"]["FWD_THRUST"] = 0.05;

        if (!in["SHIP"]["EF_THRUST"] || in["SHIP"]["REV_THRUST"].type() != INT
            || in["SHIP"]["REV_THRUST"].type() != FLOAT)
            out["SHIP"]["REV_THRUST"] = -0.05;

        if (!in["SHIP"]["TURN_RATE"] || in["SHIP"]["TURN_RATE"].type() != INT
            || in["SHIP"]["TURN_RATE"].type() != FLOAT)
            out["SHIP"]["TURN_RATE"] = 2.01;

        ////////////////////////////////////////////
        if (in["SHIP"]["COLOR"].type() == ARRAY)
        {
            if (in["SHIP"]["COLOR"][0].as_int() > 255
                || in["SHIP"]["COLOR"][0].as_int() < 0)
                out["SHIP"]["COLOR"][0] = 255;
            if (in["SHIP"]["COLOR"][1].as_int() > 255
                || in["SHIP"]["COLOR"][1].as_int() < 0)
                out["SHIP"]["COLOR"][1] = 0;
            if (in["SHIP"]["COLOR"][2].as_int() > 255
                || in["SHIP"]["COLOR"][2].as_int() < 0)
                out["SHIP"]["COLOR"][2] = 0;
        }
        else
        {
            Array color;
            color.push_back(255);
            color.push_back(0);
            color.push_back(0);
            out["SHIP"]["COLOR"] = color;
        }
/////////////////////////////////////////////////////
        if (!in["SHIP"]["OUTLINE_PX"] || in["SHIP"]["OUTLINE_PX"].type() != INT)
            out["SHIP"]["OUTLINE_PX"] = 1;
    }
    else
    {
        Object ship;
        ship["SPEED"] = 5.01;
        ship["FWD_THRUST"] = 0.05;
        ship["REV_THRUST"] = -0.05;
        ship["TURN_RATE"] = 2.01;
        Array color;
        color.push_back(255);
        color.push_back(0);
        color.push_back(0);
        ship["COLOR"] = color;
        ship["OUTLINE_PX"] = 1;
        out["SHIP"] = ship;
    }

    //write the changes
    cfgFile.open(path.c_str());
    cfgFile << out;
    cfgFile.close();

    return parse_file(path.c_str());
}

int main()
{

    cout << configInit();
}

File: StroidCfgPy/src/StroidCfg.py
#!/usr/bin/env python3.3

import os  #gives access to environment variables
import json  #builtin JSON library

class Config:
    _instance = None
    _path = os.environ['HOME'] + '/StroidConfig.json'
    #########################################################
    ##@fn _validateColor(color, defR, defG, defB)
    #@brief Checks a color to make sure it is within range and applies
    #default values if it isn't.
    #@param obj The JSON object with an Array called COLOR on which to apply the validation.
    #@param default The default RGB color, as a list (each value 0-255)
    #@return The validated JSON array.
    #
    #Like before, all objects must exist beforehand. Nonexistance of an object will
    #raise a KeyError exception. This can be changed, I just do not have the time to do so.
    def _validateColor(self, obj, default):
        tempColor = self.cfg[obj]['COLOR']
        
        for i in range(3):
            try:
                if (type(self.cfg[obj]['COLOR'][i]) is not int or
                    self.cfg[obj]['COLOR'][i] > 255 or
                    self.cfg[obj]['COLOR'][i] < 0):
                    tempColor[i] = default[i]
            except KeyError:
                tempColor[i] = default[i]
        return tempColor
    ##########################################################
    ##@fn _validateValues()
    #@brief Check all of the values in the configuration file
    #(and create them if they don't exist).
    def _validateValues(self):
        #//all if statements make sure that the property exists
        #//and that the type is correct
        
        #check frame rate
        item = 'FRAME_RATE'
        try:
            if type(self.cfg[item]) is not int or self.cfg[item] < 1:
                self.cfg[item] = 120
        except KeyError:
            self.cfg[item] = 120
            
        #check laser object
        obj = 'LASER'
        
        item = 'PULSE_LIFE'
        try:
            if type(self.cfg[obj][item]) is not int or self.cfg[obj][item] < 0:
                self.cfg[obj][item] = 100
        except KeyError:
            self.cfg[obj][item] = 100
            
        item = 'COLOR'
        self.cfg[obj][item] = self._validateColor(obj, [255, 255, 255])
        
        item = 'EDGE_DEATH'
        try:
            if type(self.cfg[obj][item]) is not bool:
                self.cfg[obj][item] = True
        except KeyError:
            self.cfg[obj][item] = True
            
        #check ship
        obj = 'SHIP'
        
        item = 'SPEED'
        try:
            if ((type(self.cfg[obj][item]) is not int and
                type(self.cfg[obj][item] is not float)) or
                self.cfg[obj][item] < 0):
                self.cfg[obj][item] = 5.01
        except KeyError:
            self.cfg[obj][item] = 5.01
        
        item = 'THRUST_FWD'
        try:
            if (type(self.cfg[obj][item]) is not int and
                type(self.cfg[obj][item] is not float)):
                self.cfg[obj][item] = .05
        except KeyError:
            self.cfg[obj][item] = .05
        item = 'THRUST_REV'
        try:
            if (type(self.cfg[obj][item]) is not int and
                type(self.cfg[obj][item] is not float)):
                self.cfg[obj][item] = -.05
        except KeyError:
            self.cfg[obj][item] = -.05
        
        item = 'TURN_RATE'
        try:
            if (type(self.cfg[obj][item]) is not int and
                type(self.cfg[obj][item] is not float)):
                self.cfg[obj][item] = 2.01
        except KeyError:
            self.cfg[obj][item] = 2.01
        
        item = 'OUTLINE_PX'
        try:
            if (type(self.cfg[obj][item]) is not int or 
                self.cfg[obj][item] < 0):
                self.cfg[obj][item] = 1
        except KeyError:
            self.cfg[obj][item] = 1
            
        item = 'COLOR'
        self.cfg[obj][item] = self._validateColor(obj, [255, 0, 0])
        
        #check asteroid
        obj = 'STROID'
        
        item = 'START_NUM'
        try:
            if (type(self.cfg[obj][item]) is not int or
                self.cfg[obj][item] < 1):
                self.cfg[obj][item] = 10
        except KeyError:
            self.cfg[obj][item] = 10
            
        item = 'SPAWN_RATE'
        try:
            if (type(self.cfg[obj][item]) is not int or
                self.cfg[obj][item] < 1):
                self.cfg[obj][item] = 100
        except KeyError:
            self.cfg[obj][item] = 100
            
        item = 'MAX_SPEED'
        try:
            if (type(self.cfg[obj][item]) is not int and
                type(self.cfg[obj][item] is not float)):
                self.cfg[obj][item] = .5
        except KeyError:
            self.cfg[obj][item] = .5
         
        ################MAKE THIS SECTION WORK CORRECTLY##########   
        item = 'SIZE_MAX'
        try:
            if (type(self.cfg[obj][item]) is not int or
                self.cfg[obj][item] < 0):
                self.cfg[obj][item] = 30
        except KeyError:
            self.cfg[obj][item] = 30
            
        item = 'SIZE_MIN'
        try:
            if (type(self.cfg[obj][item]) is not int or
                self.cfg[obj][item] < 0 or
                self.cfg[obj][item] > self.cfg[obj]['SIZE_MAX']):
                self.cfg[obj][item] = 10
        except KeyError:
            self.cfg[obj][item] = 10
            
        item = 'COLOR_RAND'
        try:
            if type(self.cfg[obj][item]) is not bool:
                self.cfg[obj][item] = True
        except KeyError:
            self.cfg[obj][item] = True
            
        item = 'COLOR'
        self.cfg[obj][item] = self._validateColor(obj, [255, 255, 255])
    #####################################################################
    def __init__(self):
        try:
            cfgFile = open(self._path)
        except FileNotFoundError:
            cfgFile = open(self._path, 'w')
            cfgStr = '{"LASER": {"EDGE_DEATH": true, "COLOR": [255, 255, 255], "PULSE_LIFE": 100}, "FRAME_RATE": 120, "STROID": {"START_NUM": 10, "SPAWN_RATE": 100, "COLOR": [255, 255, 255], "MAX_SPEED": 0.5, "SIZE_MIN": 10, "SIZE_MAX": 30, "COLOR_RAND": true}, "SHIP": {"COLOR": [255, 0, 0], "THRUST_FWD": 0.05, "SPEED": 5.01, "TURN_RATE": 2.01, "OUTLINE_PX": 1, "THRUST_REV": -0.05}}'
            j = json.loads(cfgStr)
            json.dump(j, cfgFile, indent=4)
            cfgFile.close()
        self.cfg = json.load(open(self._path))
        self._validateValues()
        cfgFile = open(self._path, 'w')
        json.dump(self.cfg, cfgFile, indent=4, sort_keys=True)
        cfgFile.close()
    ###################################################################
        #overrides function that is run when a new instance is created
        def __new__(cls, *args, **kwargs):
            if not cls._instance:
                cls._instance = super(Config, cls).__new__(cls, *args, **kwargs)
            return cls._instance
        
c = Config()
